lib/gcs_signed_url/crypto.ex

```diff
@@ -34,14 +36,59 @@ defmodule GcsSignedUrl.Crypto do
       "1fad6186e41f577a37f56589..."
   """
   @spec sign64(String.t(), Client.t()) :: String.t()
-  def sign64(string, client) do
+  def sign64(string_to_sign, client) do
     private_key = Client.get_decoded_private_key(client)
 
-    string
+    string_to_sign
     |> :public_key.sign(:sha256, private_key, rsa_padding: :rsa_pkcs1_padding)
     |> Base.encode64()
   end
 
+  @doc """
+  Signs the given string via the signBlob REST API
+  (see https://cloud.google.com/iam/docs/reference/rest/v1/projects.serviceAccounts/signBlob)
+
+  ## Examples
+
+      iex> GcsSignedUrl.Crypto.sign_via_api("foo", %GcsSignedUrl.OAuthConfig{access_token: "..."})
+      {:ok, "1fad6186e41f577a37f56589..."}
+  """
+  @spec sign_via_api(String.t(), SignBlob.OAuthConfig.t()) :: String.t()
```

Comment:
I think a code like this is hard to follow. I would go with something like this:
```elixir
  def sign_via_api(string_to_sign, %SignBlob.OAuthConfig{} = oauth_config) do
    with response <- do_post_request(string_to_sign, oauth_config),
         {:ok, body} <- Jason.decode(response.body) do
      {:ok, body["signedBlob"]}
    end
  end

  defp do_post_request(string_to_sign, oauth_config) do
    payload = Base.encode64(string_to_sign)

    @sign_blob_http.post(
      oauth_config.service_account,
      %{payload: payload},
      Authorization: "Bearer #{oauth_config.access_token}"
    )
  end
```

As you can see I didn't pattern match on `%HTTPoison.Response` because if you want to change the library you have to change the code. 
Also you can move error mapping in other place, basically where you call this function

```elixir
with {:ok, signature} <- GcsSignedUrl.Crypto.sign_via_api(...) do
   ...
else 
   error -> format_error(error)
end

Reply:
Okay I agree, the function was a bit overladed. I extracted the error formatter. But what speaks against pattern matching in the function declaration?

Reply:
I don't have any problem If you can make function head to fit in one line. I found very hard to read when you do something like this:
```elxiir
  def sign_via_api(
        string_to_sign,
        %Module1{
          some_param1: param1,
          some_param2: param2
        },
        %Module{
          some_param3: param3,
          some_param4: param4
        }
      ) do
      ...
  end
```